### PR TITLE
[.Net] Support tools for AnthropicClient and AnthropicAgent  

### PR DESCRIPTION
**Pull request by DavidLuong98**
## Why are these changes needed?
Support tools for anthropic client.

- Update `AnthropicClient`  to accept tools as an input
-  Update `AnthropicMessageConnector` to process tools
- Wrote a few UT's for client and agent

## Related issue number
Closes [#2771](https://github.com/microsoft/autogen/issues/2771)

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

File: dotnet/src/AutoGen.Anthropic/DTO/Tool.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public string? Description { get; set; }
+
+    [JsonPropertyName("input_schema")]
+    public InputSchema? InputSchema { get; set; }
+}
+
+public class InputSchema
+{
+    [JsonPropertyName("type")]
+    public string? _Type { get; set; }
```
Comment (LittleLittleCloud) [inline dotnet/src/AutoGen.Anthropic/DTO/Tool.cs:24]: why not using `Type` as the property name here

File: dotnet/test/AutoGen.Anthropic.Tests/AnthropicClientAgentTest.cs [hunk 2]
```diff
@@ -105,4 +105,47 @@ public async Task AnthropicAgentTestImageMessageAsync()
...
+            Description = "Get the current weather",
+            InputSchema = new InputSchema
+            {
+                _Type = "object",
+                Properties = new Dictionary<string, SchemaProperty>
+                {
+                    { "city", new SchemaProperty { _Type = "string", Description = "The name of the city" }},
+                    { "date", new SchemaProperty { _Type = "string", Description = "date of the day" }}
+                },
+                Required = new List<string> { "ticker" }
```
Comment (LittleLittleCloud) [inline dotnet/test/AutoGen.Anthropic.Tests/AnthropicClientAgentTest.cs:126]: "ticker"?

File: dotnet/src/AutoGen.Anthropic/Middleware/AnthropicMessageConnector.cs [hunk 3]
```diff
@@ -85,20 +85,46 @@ private async Task<IEnumerable<IMessage>> ProcessMessageAsync(IEnumerable<IMessa
...
-
     private IMessage PostProcessMessage(ChatCompletionResponse response, IAgent from)
     {
         if (response.Content is null)
         {
             throw new ArgumentNullException(nameof(response.Content));
         }
 
-        if (response.Content.Count != 1)
+        // When expecting a tool call, sometimes the response will contain two messages, one chat and one tool.
```
Comment (LittleLittleCloud) [inline dotnet/src/AutoGen.Anthropic/Middleware/AnthropicMessageConnector.cs:95]: This middleware also needs to process the following input message types to work with other function call agents
- ToolCallMessage
- ToolCallResultMessage
- ToolCallAggregateMessage

You can refer to `OpenAI` or `Gemini` on how those messages being converted

**Comment by LittleLittleCloud** [comment]
Can you also update samples on how to use function call in `AutoGen.Anthropic.Samples` project.

maybe accomodate from one of these samples in anthropic api document
- https://docs.anthropic.com/en/docs/tool-use-examples

File: dotnet/src/AutoGen.Anthropic/DTO/Tool.cs (hunk 1 above)
Comment (DavidLuong98) [reply dotnet/src/AutoGen.Anthropic/DTO/Tool.cs]: Updated.

I was too use to naming it `_Type` because of NAOT deserialization. When the fields are named after a reserved keyword, the generated files will cause an error. but for this case, i did not need to prefix it with an underscore. :)

File: dotnet/src/AutoGen.Anthropic/Middleware/AnthropicMessageConnector.cs (hunk 3 above)
Comment (DavidLuong98) [reply dotnet/src/AutoGen.Anthropic/Middleware/AnthropicMessageConnector.cs:101]: Could you give me feedback on this ? i followed the pattern of `OpenAi` and `Gemini`. Wrote Ut's that invoke these methods and they all pass. Looking for second opinion if i did this correctly. Hopefully the changes do not hurt your eyes 😁

Comment (LittleLittleCloud) [reply dotnet/src/AutoGen.Anthropic/Middleware/AnthropicMessageConnector.cs:101]: Sorry for the late reply, was afk in the past few days,,

These message types (`ToolCallMessage`, `ToolCallResultMessage` and `ToolCallAggregateMessage`) also need to be processed so that the agent can process function call message types correctly when they are in the chat history.

For example, considering the following message history
```
TextMessage("What's the weather in seattle", from: user)
ToolCallAggregateMessage(ToolCallMessage("getWeather"), ToolCallResultMessage("the weather in seattle is sunny"))
```

To perform a multi-turn function call where the function call result needs to be sent to LLM for further response, the agent also needs to support `ToolCallAggregateMessage` as input.

Similiary, `ToolCallMessage` and `ToolCallResultMessage` needs to be supported as well.

Comment (DavidLuong98) [reply dotnet/src/AutoGen.Anthropic/Middleware/AnthropicMessageConnector.cs:112]: sorry for the delay, been a little busy with regular day job  😆
`
I added support for `ToolCallMessage`, `ToolCallAggregateMessage`, and `ToolCallResultMessage` 

Let me know if this looks good to you, added unit tests as well.

File: dotnet/sample/AutoGen.Anthropic.Samples/Create_Antrhopic_Agent.cs [hunk 4]
```diff
@@ -7,7 +7,7 @@
 
 namespace AutoGen.Anthropic.Samples;
 
-public static class AnthropicSamples
+public static class Create_Antrhopic_Agent
```
Comment (LittleLittleCloud) [inline dotnet/sample/AutoGen.Anthropic.Samples/Create_Antrhopic_Agent.cs:10]: `Antrhopic` -> `Anthropic`

File: dotnet/src/AutoGen.Anthropic/AnthropicClient.cs [hunk 5]
```diff
@@ -61,33 +61,75 @@ public AnthropicClient(HttpClient httpClient, string baseUrl, string apiKey)
         using var reader = new StreamReader(await httpResponseMessage.Content.ReadAsStreamAsync());
 
         var currentEvent = new SseEvent();
+
         while (await reader.ReadLineAsync() is { } line)
         {
             if (!string.IsNullOrEmpty(line))
             {
-                currentEvent.Data = line.Substring("data:".Length).Trim();
+                if (line.StartsWith("event:"))
```
Comment (LittleLittleCloud) [inline dotnet/src/AutoGen.Anthropic/AnthropicClient.cs:69]: The data payload of `Server side event` should always start with `data:`. Is that something that is anthropic specifc?

File: dotnet/sample/AutoGen.BasicSamples/CodeSnippet/TypeSafeFunctionCallCodeSnippet.cs [hunk 6]
```diff
@@ -11,6 +12,15 @@
 #region weather_report
 public partial class TypeSafeFunctionCall
 {
+    private class GetWeatherSchema
```
Comment (LittleLittleCloud) [inline dotnet/sample/AutoGen.BasicSamples/CodeSnippet/TypeSafeFunctionCallCodeSnippet.cs:15]: Is this change necessary?

File: dotnet/src/AutoGen.Anthropic/AnthropicClient.cs (hunk 5 above)
Comment (DavidLuong98) [reply dotnet/src/AutoGen.Anthropic/AnthropicClient.cs:69]: Yes this is anthropic specific. This is an example response from docs

```
event: message_start
data: {"type": "message_start", "message": {"id": "msg_1nZdL29xx5MUA1yADyHTEsnR8uuvGzszyY", "type": "message", "role": "assistant", "content": [], "model": "claude-3-5-sonnet-20240620", "stop_reason": null, "stop_sequence": null, "usage": {"input_tokens": 25, "output_tokens": 1}}}

event: content_block_start
data: {"type": "content_block_start", "index": 0, "content_block": {"type": "text", "text": ""}}

event: ping
data: {"type": "ping"}

event: content_block_delta
data: {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "Hello"}}

event: content_block_delta
data: {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "!"}}

event: content_block_stop
data: {"type": "content_block_stop", "index": 0}

event: message_delta
data: {"type": "message_delta", "delta": {"stop_reason": "end_turn", "stop_sequence":null}, "usage": {"output_tokens": 15}}

event: message_stop
data: {"type": "message_stop"}
```

File: dotnet/sample/AutoGen.BasicSamples/CodeSnippet/TypeSafeFunctionCallCodeSnippet.cs (hunk 6 above)
Comment (DavidLuong98) [reply dotnet/sample/AutoGen.BasicSamples/CodeSnippet/TypeSafeFunctionCallCodeSnippet.cs:15]: I made this change for one of the unit test, `AnthropicAgentFunctionCallMessageTest()` . Do you prefer if i create my own FunctionCall in `Anthropic.Tests` instead?

Comment (LittleLittleCloud) [reply dotnet/sample/AutoGen.BasicSamples/CodeSnippet/TypeSafeFunctionCallCodeSnippet.cs:15]: Yes I feel like it would be better. This code snippet is used in website so changing it will also change the code example in the tutorial as well.

File: dotnet/src/AutoGen.Anthropic/Agent/AnthropicClientAgent.cs [hunk 7]
```diff
@@ -16,21 +17,24 @@ public class AnthropicClientAgent : IStreamingAgent
...
 
     public AnthropicClientAgent(
         AnthropicClient anthropicClient,
         string name,
         string modelName,
         string systemMessage = "You are a helpful AI assistant",
         decimal temperature = 0.7m,
-        int maxTokens = 1024)
+        int maxTokens = 1024,
+        Tool[]? tools = null)
```
[the remark this answers is not in the document]
Comment (DavidLuong98) [reply dotnet/src/AutoGen.Anthropic/Agent/AnthropicClientAgent.cs]: Yes, good idea. I added this. Resolving this comment

File: dotnet/sample/AutoGen.BasicSamples/CodeSnippet/TypeSafeFunctionCallCodeSnippet.cs (hunk 6 above)
Comment (DavidLuong98) [reply dotnet/sample/AutoGen.BasicSamples/CodeSnippet/TypeSafeFunctionCallCodeSnippet.cs]: Yes. i reverted this usage and added this example under Anthropic.Test proj instead

**Review by LittleLittleCloud — APPROVED** [review]

**Comment by LittleLittleCloud** [comment]
@DavidLuong98 Is this good to merge in

**Comment by DavidLuong98** [comment]
@LittleLittleCloud  yes
